xdsl/dialects/riscv.py

```diff
@@ -3016,6 +3017,68 @@ def assembly_line_args(self) -> tuple[AssemblyInstructionArg, ...]:
         return self.rd, self.rs1, self.rs2
 
 
+class RdRsRsFloatFloatIntegerOperationWithFastMath(RISCVInstruction, ABC):
+    """
+    A base class for RISC-V operations that have two source floating-point
+    registers with an integer destination register, and can be annotated with fastmath flags.
+
+    This is called R-Type in the RISC-V specification.
+    """
+
+    rd = result_def(IntRegisterType)
+    rs1 = operand_def(FloatRegisterType)
+    rs2 = operand_def(FloatRegisterType)
+    fastmath = prop_def(FastMathFlagsAttr)
+
+    def __init__(
+        self,
+        rs1: Operation | SSAValue,
+        rs2: Operation | SSAValue,
+        *,
+        rd: IntRegisterType | str | None = None,
+        fastmath: FastMathFlagsAttr = FastMathFlagsAttr("none"),
+        comment: str | StringAttr | None = None,
+    ):
+        if rd is None:
+            rd = IntRegisterType.unallocated()
+        elif isinstance(rd, str):
+            rd = IntRegisterType(rd)
+        if isinstance(comment, str):
+            comment = StringAttr(comment)
+
+        super().__init__(
+            operands=[rs1, rs2],
+            properties={
+                "fastmath": fastmath,
+            },
+            attributes={
+                "comment": comment,
+            },
+            result_types=[rd],
+        )
+        self.fastmath = fastmath
+
+    def assembly_line_args(self) -> tuple[AssemblyInstructionArg, ...]:
+        return self.rd, self.rs1, self.rs2
+
+    @classmethod
+    def custom_parse_attributes(cls, parser: Parser) -> dict[str, Attribute]:
+        attributes = dict[str, Attribute]()
+        fast = FastMathFlagsAttr("none")
+        if parser.parse_optional_keyword("fastmath") is not None:
+            fast = FastMathFlagsAttr(FastMathFlagsAttr.parse_parameter(parser))
+            if fast != FastMathFlagsAttr("none"):
+                attributes["fastmath"] = fast
+        cls.fastmath = fast
```

Comment:
This sets fastmath on the class directly, not on the instance. What happens without this line? I would expect that there is some magic to convert attributes to properties, but if not we'll need to add this to our infrastructure. I'll pull this branch locally and see what's up.

Reply:
OK so the current parsing infrastructure only handles attributes. I would recommend changing fastmath to a non-optional `attr_def`, and not dealing with the whole property infrastructure side of things

Reply:
Eventually we'll want to migrate the riscv dialect to use properties instead of attributes, but that's not critical

Reply:
it is now using `attr_def`